File: interview-questions/2018-06-03/solution.py
def has_pair_with_sum(arr, sum):
    comp = {}
    for v in arr:
        if v in comp.keys():
            return True
        comp[sum-v] = None
    return False

if __name__ == "__main__":
    input1 = [1,2,4,9]
    print(has_pair_with_sum(input1, 8))

    input2 = [1,9,3,6,1,7,9]
    print(has_pair_with_sum(input2, 8))

File: interview-questions/2018-06-03/solution.cc
/*
 * Solution by Edgar Duéñez-Guzmán
 * https://youtu.be/XKu_SEDAykw?t=14m53s
 */
#include <iostream>
#include <unordered_set>
#include <vector>

using namespace std;

bool HasPairWithSum(const vector<int>& data, int sum) {
        unordered_set<int> comp; // additive complements
        for (int value : data) {
                if (comp.find(value) != comp.end()) return true;
                comp.insert(sum-value);
        }
        return false;
}

int main() {
        cout << HasPairWithSum({1,2,4,9}, 8) << endl;
        cout << HasPairWithSum({1,2,4,4}, 8) << endl;
}
